File: common/src/main/java/com/pluxity/permission/Permission.java
package com.pluxity.permission;

import com.pluxity.global.entity.BaseEntity;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Table(name = "permission")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Permission extends BaseEntity {

    @Id
    @GeneratedValue(strategy = jakarta.persistence.GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    private PermissionGroup permissionGroup;

    @Column(nullable = false)
    private String resourceName;

    @Column(nullable = false)
    private String resourceId;

    @Builder
    public Permission(String resourceName, String resourceId, PermissionGroup permissionGroup) {
        this.resourceName = resourceName;
        this.resourceId = resourceId;
        this.permissionGroup = permissionGroup;
    }

    public void changeResourceName(String resourceName) {
        this.resourceName = resourceName;
    }

    public void changeResourceId(String resourceId) {
        this.resourceId = resourceId;
    }

    public boolean matches(String resourceName, String resourceId) {
        return this.resourceName.equalsIgnoreCase(resourceName) && this.resourceId.equals(resourceId);
    }

    protected void changePermissionGroup(PermissionGroup permissionGroup) {
        if (this.permissionGroup != null) {
            this.permissionGroup.getPermissions().remove(this);
        }
        this.permissionGroup = permissionGroup;
        if (permissionGroup != null && !permissionGroup.getPermissions().contains(this)) {
            permissionGroup.getPermissions().add(this);
        }
    }

    protected void clearPermissionGroup() {
        this.permissionGroup = null;
    }
}

File: common/src/main/kotlin/com/pluxity/permission/Permission.kt
package com.pluxity.permission

import com.pluxity.global.entity.IdentityIdEntity
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.FetchType
import jakarta.persistence.ManyToOne
import jakarta.persistence.Table

@Entity
@Table(name = "permission")
class Permission(
    @Column(nullable = false)
    var resourceName: String,
    @Column(nullable = false)
    var resourceId: String,
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var level: PermissionLevel = PermissionLevel.READ,
    @ManyToOne(fetch = FetchType.LAZY)
    var permissionGroup: PermissionGroup? = null,
) : IdentityIdEntity() {
    fun allows(
        resourceName: String,
        resourceId: String,
        requiredLevel: PermissionLevel,
    ): Boolean =
        this.resourceName.equals(resourceName, ignoreCase = true) &&
            this.resourceId == resourceId &&
            this.level.allows(requiredLevel)

    fun changePermissionGroup(permissionGroup: PermissionGroup?) {
        this.permissionGroup?.permissions?.remove(this)
        this.permissionGroup = permissionGroup
        permissionGroup?.permissions?.let { permissions ->
            if (!permissions.contains(this)) {
                permissions.add(this)
            }
        }
    }

    fun clearPermissionGroup() {
        this.permissionGroup = null
    }
}
